Zend/zend_enum.c

```diff
@@ -0,0 +1,338 @@
+
+	char* forbidden_methods[] = {
+		"__sleep",
+		"__wakeup",
+		"__set_state",
+	};
+
+	uint32_t forbidden_methods_length = sizeof(forbidden_methods) / sizeof(forbidden_methods[0]);
+	for (uint32_t i = 0; i < forbidden_methods_length; ++i) {
+		char *forbidden_method = forbidden_methods[i];
+
+		if (zend_hash_str_find_ptr(&ce->function_table, forbidden_method, strlen(forbidden_method))) {
+			zend_error_noreturn(E_COMPILE_ERROR, "Enum may not include magic method %s", forbidden_method);
+		}
+	}
+}
+
+void zend_verify_enum(zend_class_entry *ce)
+{
+	zend_verify_enum_properties(ce);
+	zend_verify_enum_magic_methods(ce);
+}
+
+static zval *zend_enum_read_property(zend_object *zobj, zend_string *name, int type, void **cache_slot, zval *rv) /* {{{ */
+{
+	if (type == BP_VAR_W) {
+		zend_throw_error(NULL, "Cannot acquire reference to property %s::$%s", ZSTR_VAL(zobj->ce->name), ZSTR_VAL(name));
+		return &EG(uninitialized_zval);
+	}
+
+	return zend_std_read_property(zobj, name, type, cache_slot, rv);
+}
+
+ZEND_COLD zval *zend_enum_write_property(zend_object *object, zend_string *member, zval *value, void **cache_slot)
+{
+	ZEND_ENUM_PROPERTY_ERROR();
+	return &EG(uninitialized_zval);
+}
+
+static ZEND_COLD void zend_enum_unset_property(zend_object *object, zend_string *member, void **cache_slot)
+{
+	ZEND_ENUM_PROPERTY_ERROR();
+}
+
+ZEND_API zval *zend_enum_get_property_ptr_ptr(zend_object *zobj, zend_string *name, int type, void **cache_slot)
+{
+	return NULL;
+}
+
+int zend_enum_compare_objects(zval *o1, zval *o2)
+{
+	if (Z_TYPE_P(o1) != IS_OBJECT || Z_TYPE_P(o2) != IS_OBJECT) {
+		return ZEND_UNCOMPARABLE;
+	}
+
+	return Z_OBJ_P(o1) == Z_OBJ_P(o2) ? 0 : 1;
+}
+
+void zend_register_enum_ce(void)
+{
+	memcpy(&enum_handlers, &std_object_handlers, sizeof(zend_object_handlers));
+	enum_handlers.read_property = zend_enum_read_property;
+	enum_handlers.write_property = zend_enum_write_property;
+	enum_handlers.unset_property = zend_enum_unset_property;
+	enum_handlers.get_property_ptr_ptr = zend_enum_get_property_ptr_ptr;
+	enum_handlers.clone_obj = NULL;
+	enum_handlers.compare = zend_enum_compare_objects;
+}
+
+void zend_enum_add_interfaces(zend_class_entry *ce)
+{
+	uint32_t num_interfaces_before = ce->num_interfaces;
+
+	ce->num_interfaces++;
+	if (ce->enum_scalar_type != IS_UNDEF) {
+		ce->num_interfaces++;
+	}
+
+	ce->interface_names = erealloc(ce->interface_names, sizeof(zend_class_name) * ce->num_interfaces);
+
+	ce->interface_names[num_interfaces_before].name = zend_string_init("UnitEnum", sizeof("UnitEnum") - 1, 0);
+	ce->interface_names[num_interfaces_before].lc_name = zend_string_init("unitenum", sizeof("unitenum") - 1, 0);
+
+	if (ce->enum_scalar_type != IS_UNDEF) {
+		ce->interface_names[num_interfaces_before + 1].name = zend_string_init("ScalarEnum", sizeof("ScalarEnum") - 1, 0);
+		ce->interface_names[num_interfaces_before + 1].lc_name = zend_string_init("scalarenum", sizeof("scalarenum") - 1, 0);	
+	}
```

Comment:
Is it possible for an enum to manually implement the UnitEnum/ScalarEnum interfaces?

Reply:
No, the error message currently looks something like this:

> Class Foo cannot implement previously implemented interface UnitEnum

The `class` part could be confusing but I'm afraid if we start changing it here we'll have to account for it all over the place.